Add Integration tests (#11)
AdaGold/swap-meet: README.md

We will use a Test Driven Development programming workflow to work on this project. Notice the Red-Green-Refactor steps in the workflow steps outlined below.

1. When you want to begin work on this project, ensure that your virtual environment is activated:

```bash
$ source venv/bin/activate
```

2. Check the `tests/unit_tests` folder, and find the test file you want to run

   - In that test file, read through each test case
      - If it is incomplete, complete the test. **There are incomplete tests for Waves 01, 02, and 06**. 
        - *Is this a nominal or edge case?*
        - *What type of input do we need to test this case?*
        - *What is the expected output for the given input?*
   - Remove the lines that contain `@pytest.mark.skip()` for the test(s) you want to run.

3. Run the test(s)! (RED)
    - *See the [Details About How to Run Tests](#details-about-how-to-run-tests) section below for more information on how to run test(s).*

```bash
# Must be in activated virtual environment in the project-root directory
$ pytest
```

4. Read through the test failure, and understand why the failure is happening. Confirm your findings with a classmate. 

>     - If it is a test you wrote, consider whether you are actually testing what you intend to test. Does the test need modification?

Reviewer: Feel free to ignore, I just had the thought that this tweak might focus on what the code is doing.

```suggestion
    - If it is a test you wrote, consider whether the code is actually testing what you intend to test. Does the test need modification?
```

5. Make a plan to implement code to pass the test.

6. Write code to pass the test (GREEN).

7. Re-run the test(s).

8. Repeat steps 3-7 until that test passes!

9. Consider looking for opportunities to improve your code (REFACTOR)

> 10. Repeats steps 2-8 until you have finished all tests in the file.

Reviewer: Couple small things:
- Should it be steps 2-9 to include the refactor step?
- I think "Repeats" should be "Repeat"

11. Begin using the test file of the next wave!

> 12. When you are finished working for the day, deactivate your environment with deactivate or closing the Terminal tab/window

Reviewer: Small grammar suggestion

```suggestion
12. When you are finished working for the day, deactivate your environment with deactivate or by closing the Terminal tab/window
```

```bash
$ deactivate
```

> Finally, at submission time, **no matter where you are**, submit the project via Learn.

Reviewer: I like adding the emphasis here to turn in the project wherever it may be.

This will let us give feedback on what you've finished so that you can be better prepared for the next project.

## Integration Tests

> We provided integration tests for this project. Integration tests verify that individual units of a piece of software are working together correctly.

Reviewer: Is there a way to reword this that helps students understand how the integration tests are distinct from unit tests/the other tests in the project? The paragraph below definitely implies that they are more involved than unit tests, but my thinking is more clarity could be helpful.

Author: I took a stab at this reorganization. Would love feedback on this!

The integration tests for this project are in the directory `tests/integration_tests`, and the tests have the decorator `@pytest.mark.integration_test` provided by the `pytest-integration` package. Marking these tests as integration tests makes them run after the unit tests. The isolated nature and specficity of *unit tests* make them a much better tool for debugging. Hence, we want to run the unit tests before the integration tests.

> The integration tests provided in this project do not cover all the test cases verified by the unit tests. They verify that the functions work together when invoked correctly. We could classify them as the subset of Integration Tests called *Smoke Tests*, tests that check basic functionality of the application.

Reviewer: This paragraph really gets at what I meant with my prior comment on line 37. Would it make sense to move this up as part of that first paragraph that introduces what the integration tests are?

Author: Good call. I took a stab at reorganizing.

Reviewer: I think I'm stuck on the sentence "Integration tests verify that individual units of a piece of software are working together correctly." To me, it doesn't add as much clarity as I would like.  

I was thinking something that compared them directly to unit tests like:
> While unit tests are small, fast, and should cover most of our program's behavior down to individual functions, integration tests verify that the various pieces of a program are working together correctly.

Reviewer 2: I like Kelsey's suggestion. I think specifically, using the word 'units' in the definition of integration tests might undermine the contrast we're trying to draw between unit and integration tests.

Author: great! I incorporated this.

For more on different types of software testing, [here is a helpful resource](https://www.atlassian.com/continuous-delivery/software-testing/types-of-software-testing). 

## Code Coverage

> It is good practice to check code coverage. Code coverage is a term that is used to describe how much application code is executed when a particular test suite is run. A test suite with a high percentage of coverage is likely to be testing more throughly and have fewer bugs. A code coverage tool can partner with our testing suite to give us a report illustrating the coverage of our tests.

Reviewer: I was thinking it might be helpful to define code coverage first and provide a little more "why" in the good practice sentence. 

```suggestion
Code coverage is a term used to describe how much application code is executed when a particular test suite is run. It is a good practice to check our code coverage, to understand how much of our code is exercised by tests vs how much is still untested. A test suite with a high percentage of coverage is likely to be testing more throughly and have fewer bugs. A code coverage tool can partner with our testing suite to give us a report illustrating the coverage of our tests.
```

Given that Ada provided all tests cases in this project, we should anticipate high code coverage.

Review the [code coverage exercise](https://github.com/adaGold/code-coverage-exercise) on how to use `pytest-cov` to generate a code coverage report. We will need to change the directory where the application code is located from `student` to `swap_meet`.  

`pytest --cov=swap_meet --cov-report html --cov-report term`

*Note: Code coverage is disabled for integration tests, since unit tests should cover all the code.* [source](https://pypi.org/project/pytest-integration/#:~:text=Overview,integration%20tests%20are%20not%20run)

## Details About How to Run Tests

All the commands described below should be run from the project-root directory `swap_meet`. Note that the project-root directory is the repository `swap-meet`. It is distinct from the directory `swap_meet` that contains the source code.

To run all unskipped tests that exist in this project with:

```bash
# Must be in activated virtual environment
$ pytest
```

> To see any `print` statements print to the console, add `-s` to the end of any `pytest` command:

Reviewer: Minor suggestion, not necessary, but this reads a little more naturally to me:

```suggestion
To see any `print` statement's output in the console, add `-s` to the end of any `pytest` command:
```

```bash
# Must be in activated virtual environment
$ pytest -s
```

To run all unskipped tests that exist in one file, use:

```bash
# Must be in activated virtual environment
$ pytest tests/unit_tests/test_file_name.py
```

... where `test_file_name.py` is replaced with the correct test file name.

> To run a single test by name:

Reviewer: I like the gentle reorganization here and adding in running a single test by name.

$ source venv/bin/activate
```

2. Check the `tests/unit_tests` folder, and find the test file you want to run

   - In that test file, read through each test case
      - If it is incomplete, complete the test. **There are incomplete tests for Waves 01, 02, and 06**. 
        - *Is this a nominal or edge case?*
        - *What type of input do we need to test this case?*
        - *What is the expected output for the given input?*

>    - Remove the lines that contain `@pytest.mark.skip()` for the test(s) you want to run.

Reviewer 2: It doesn't seem like `@pytest.mark.skip()` is used for any tests in this project, so could be omitted unless this is just general information included for all projects

Author: Oh! Good catch. I was thinking I'd add these to each test. Though open to leaving them off. I know they are in viewing party for each test.

Reviewer: If students won't have seen this before, would it be worth adding something that reminds them to search the project/test files for `@pytest.mark.skip()` to ensure no tests were missed before submitting the assignment?

Author: Added @pytest.mark.skip to all tests and added a note to search for this in the readme (actually viewing-party readme, which this readme now points to)

> 3. Run the test(s)! (RED)

Author: I meant to add this line to the top of the project workflow "We will use a Test Driven Development programming workflow to work on this project. Notice the Red-Green-Refactor steps in the workflow steps outlined below.", and then add (GREEN) to step 6

> 3. Run the test(s)! (RED)

Reviewer: I'm not familiar enough with our curriculum to know the answer to this, but will students have seen this notation before? I know a lot of folks will probably immediately get red is for test should be not passing here, but my instinct is this could raise confusion if they haven't definitely seen it before, even with the paragraph above.

> 3. Run the test(s)! (RED)

Author: Yea, I think you're right that it adds more confusion than value. I'll remove it here and from the viewing party README.

    - *See the [Details About How to Run Tests](#details-about-how-to-run-tests) section below for more information on how to run test(s).*

```bash
# Must be in activated virtual environment in the project-root directory
$ pytest
```

4. Read through the test failure, and understand why the failure is happening. Confirm your findings with a classmate. 
    - If it is a test you wrote, consider whether you are actually testing what you intend to test. Does the test need modification?

5. Make a plan to implement code to pass the test.

> 6. Write code to pass the test.

Author: ```suggestion
6. Write code to pass the test. (GREEN)
```
